JuliaStream.jl/src/CUDAStream.jl

```diff
@@ -0,0 +1,154 @@
+      scalar,
+      arraysize,
+    ),
+    nothing,
+  )
+end
+
+function init_arrays!(data::CuData{T}, _, init::Tuple{T,T,T}) where {T}
+  CUDA.fill!(data.a, init[1])
+  CUDA.fill!(data.b, init[2])
+  CUDA.fill!(data.c, init[3])
+end
+
+function copy!(data::CuData{T}, _) where {T}
+  function kernel(a, c)
+    i = (blockIdx().x - 1) * blockDim().x + threadIdx().x # only blockIdx starts at 1
+    @inbounds c[i] = a[i]
+    return
+  end
+  @cuda blocks = data.size ÷ TBSize threads = TBSize kernel(data.a, data.c)
+  CUDA.synchronize()
+end
+
+function mul!(data::CuData{T}, _) where {T}
+  function kernel(b, c, scalar)
+    i = (blockIdx().x - 1) * blockDim().x + threadIdx().x # only blockIdx starts at 1
+    @inbounds b[i] = scalar * c[i]
+    return
+  end
+  @cuda blocks = data.size ÷ TBSize threads = TBSize kernel(data.b, data.c, data.scalar)
+  CUDA.synchronize()
+end
+
+function add!(data::CuData{T}, _) where {T}
+  function kernel(a, b, c)
+    i = (blockIdx().x - 1) * blockDim().x + threadIdx().x # only blockIdx starts at 1
+    @inbounds c[i] = a[i] + b[i]
+    return
+  end
+  @cuda blocks = data.size ÷ TBSize threads = TBSize kernel(data.a, data.b, data.c)
+  CUDA.synchronize()
+end
+
+function triad!(data::CuData{T}, _) where {T}
+  function kernel(a, b, c, scalar)
+    i = (blockIdx().x - 1) * blockDim().x + threadIdx().x # only blockIdx starts at 1
+    @inbounds a[i] = b[i] + (scalar * c[i])
+    return
+  end
+  @cuda blocks = data.size ÷ TBSize threads = TBSize kernel(
+    data.a,
+    data.b,
+    data.c,
+    data.scalar,
+  )
+  CUDA.synchronize()
+end
+
+function nstream!(data::CuData{T}, _) where {T}
+  function kernel(a, b, c, scalar)
+    i = (blockIdx().x - 1) * blockDim().x + threadIdx().x # only blockIdx starts at 1
+    @inbounds a[i] += b[i] + scalar * c[i]
+    return
+  end
+  @cuda blocks = data.size ÷ TBSize threads = TBSize kernel(
+    data.a,
+    data.b,
+    data.c,
+    data.scalar,
+  )
+  CUDA.synchronize()
+end
+
+function dot(data::CuData{T}, _) where {T}
+  # direct port of the reduction in CUDAStream.cu 
+  function kernel(a, b, size, partial)
+    tb_sum = @cuStaticSharedMem(T, TBSize)
+    local_i = threadIdx().x
+    @inbounds tb_sum[local_i] = 0.0
+
+    # do dot first
+    i = (blockIdx().x - 1) * blockDim().x + threadIdx().x # only blockIdx starts at 1
+    while i <= size
+      @inbounds tb_sum[local_i] += a[i] * b[i]
+      i += blockDim().x * gridDim().x
+    end
+
+    # then tree reduction
+    offset = blockDim().x ÷ 2
+    while offset > 0
+      sync_threads()
+      if (local_i - 1) < offset
+        @inbounds tb_sum[local_i] += tb_sum[local_i+offset]
+      end
+      offset ÷= 2
+    end
+
+    if (local_i == 1)
+      @inbounds partial[blockIdx().x] = tb_sum[local_i]
+    end
+
+    return
+  end
+  partial_sum = CuArray{T}(undef, DotBlocks)
+  @cuda blocks = DotBlocks threads = TBSize kernel(data.a, data.b, data.size, partial_sum)
+  CUDA.synchronize()
```

Comment:
The sum below will run on the same stream as this kernel, so you don't need the synchronization.